components/ai_chat/resources/untrusted_conversation_frame/components/assistant_response/assistant_response.test.tsx

```diff
@@ -55,3 +57,93 @@ test('AssistantResponse should include expandable sources', async () => {
   links = screen.getAllByRole('link')
   expect(links).toHaveLength(8)
 })
+
+test('AssistantResponse should render memory tool events inline', async () => {
+  const mockHasMemory = jest.fn().mockResolvedValue({ exists: true })
+  const mockUIObserver = {
+    onMemoriesChanged: {
+      addListener: jest.fn().mockReturnValue('listener-id')
+    },
+    removeListener: jest.fn()
+  }
+
+  const memoryToolEvent: Mojom.ConversationEntryEvent = {
+    toolUseEvent: {
+      toolName: Mojom.MEMORY_STORAGE_TOOL_NAME,
+      id: 'memory-tool-123',
+      argumentsJson: '{"memory": "Test memory content"}',
+      output: [createTextContentBlock('')]
+    }
+  }
+
+  const events = [
+    memoryToolEvent,
+    getCompletionEvent('I will remember that.')
+  ]
+
+  render(
+    <MockContext
+      uiHandler={{
+        hasMemory: mockHasMemory
+      } as unknown as Mojom.UntrustedUIHandlerRemote}
+      uiObserver={mockUIObserver as unknown as Mojom.UntrustedUICallbackRouter}
+    >
+      <AssistantResponse
+        events={events}
+        isEntryInteractivityAllowed={false}
+        isLeoModel={true}
+        isEntryInProgress={false}
+        allowedLinks={[]}
+      />
+    </MockContext>
+  )
+
+  // Memory tool should render inline (success state)
+  await waitFor(() => {
+    expect(screen.getByTestId('memory-tool-event')).toBeInTheDocument()
+  })
+  expect(screen.getByText(/Test memory content/)).toBeInTheDocument()
+})
+
+test(
+  'AssistantResponse should render memory tool in undone state when memory ' +
+    'does not exist',
+     async () => {
+  const mockHasMemory = jest.fn().mockResolvedValue({ exists: false })
+
+  const memoryToolEvent: Mojom.ConversationEntryEvent = {
+    toolUseEvent: {
+      toolName: Mojom.MEMORY_STORAGE_TOOL_NAME,
+      id: 'memory-tool-123',
+      argumentsJson: '{"memory": "Test memory content"}',
+      output: [createTextContentBlock('')]
+    }
+  }
+
+  const events = [
+    memoryToolEvent,
+    getCompletionEvent('I will remember that.')
+  ]
+
+  render(
+    <MockContext
+      uiHandler={{
+        hasMemory: mockHasMemory
```

Comment:
do we need to use a jest mock since we're not testing how it's been called. Seems like we can just provide `hasMemory: Promise.resolve(true)`

Reply:
It seems it would fail with `Uncaught [TypeError: context.uiHandler?.hasMemory is not a function]`.